Wrap LabelSetMeasures in LabelOverlapMeasuresImageFilter

dzenanz opened the pull request:
This exposes the return value from `itk::LabelOverlapMeasuresImageFilter::GetLabelSetMeasures()` as a Python dictionary instead of `<Swig Object of type 'std::unordered_map< unsigned char,itkLabelOverlapMeasuresImageFilterIUC3::LabelSetMeasures > *'>`. This is needed for convenient use of the class from Python.

## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [x] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5

dzenanz committed:
ENH: Wrap LabelSetMeasures in LabelOverlapMeasuresImageFilter

This exposes the return value from
itk::LabelOverlapMeasuresImageFilter::GetLabelSetMeasures()
as a Python dictionary instead of
<Swig Object of type 'std::unordered_map< unsigned char,itkLabelOverlapMeasuresImageFilterIUC3::LabelSetMeasures > *'>
This is needed for convenient use of the class from Python.
The class has been un-nested to make wrapping easier/possible.

dzenanz committed:
ENH: Add a Python test

dzenanz commented:
Currently running into:
```log
3024: Loading ITKImageStatistics... done
3024: Running itkLabelOverlapMeasuresImageFilterIUC3... Running itkLabelOverlapMeasuresImageFilterIUC3... done
3024: done
3024: Traceback (most recent call last):
3024:   File "C:\Dev\ITK-git\Modules\Filtering\ImageStatistics\wrapping\test\itkLabelOverlapMeasuresImageFilterTest.py", line 33, in <module>
3024:     for label, measure in lsm.items():
3024:                           ^^^^^^^^^
3024: AttributeError: 'SwigPyObject' object has no attribute 'items'
3024: swig/python detected a memory leak of type 'std::unordered_map< unsigned char,itkLabelOverlapLabelSetMeasures,std::hash< unsigned char >,std::equal_to< unsigned char >,std::allocator< std::pair< unsigned char const,itkLabelOverlapLabelSetMeasures > > > *', no destructor found.
3024: itkTestDriver: Process exited with return value: 1
```

thewtex commented:
What is `dir(lsm)`?

dzenanz commented:
`dir(lsm)` is `['__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__int__', '__le__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'acquire', 'append', 'disown', 'next', 'own']`

thewtex commented:
@dzenanz could you please rebase and push? I am wondering if recent SWIG updates have helped.

hjmjohnson commented:
Rebased onto current `upstream/main` (was 3894 commits behind).  Both commits cherry-picked cleanly with only auto-merges in `itkLabelOverlapMeasuresImageFilter.h` and `Wrapping/Generators/Python/PyBase/pyBase.i` — no manual conflict resolution needed.  CI will re-run on the new HEAD; the SWIG-update question @thewtex raised in May 2024 should now answer itself.

hjmjohnson committed:
ENH: Add Python-friendly LabelSetMeasures accessors

Expose the per-label overlap measures to Python via two paired
accessors on itk::LabelOverlapMeasuresImageFilter:

  std::vector<LabelType> GetLabels() const
  LabelOverlapLabelSetMeasures GetMeasureForLabel(LabelType) const

Plus six explicit `Get*()` getters on itk::LabelOverlapLabelSetMeasures
(GetSource, GetTarget, GetUnion, GetIntersection, GetSourceComplement,
GetTargetComplement) so Python can read the per-label fields.

The existing C++ API (`MapType GetLabelSetMeasures()` returning
`std::unordered_map<LabelType, LabelOverlapLabelSetMeasures>`) is
preserved unchanged for performant in-process lookup; the new
accessors are additive.

Why two accessors instead of wrapping the unordered_map as a Python
dict: SWIG's %template instantiation of std::unordered_map<X,Y> does
not materialize a dict-like wrapper class when Y (the value type) is
%import-ed (rather than %include-d) into the consuming submodule.
ITK's per-submodule wrapping pipeline isolates each class as its own
SWIG run with %import-only type info from siblings, so the
%template directives in itkLabelOverlapLabelSetMeasuresInstantiations.i
were silently dropped (count of `hashmap` in the SWIG-generated .cpp
was 0).  Falling back to a vector-of-keys + per-key lookup avoids the
%template materialization issue entirely; std::vector<LabelType> is
already wrapped globally in pyBase.i for every primitive label type.

Co-Authored-By: dzenanz <dzenanz@gmail.com>

hjmjohnson committed:
COMP: Drop unused std::unordered_map SWIG infrastructure

The original PR added wrapping infrastructure to expose
`std::unordered_map<LabelType, LabelOverlapLabelSetMeasures>` as a
Python dict.  In practice the %template instantiations in
itkLabelOverlapLabelSetMeasuresInstantiations.i were silently dropped
by SWIG: the value type LabelOverlapLabelSetMeasures is %import-ed
(not %include-d) into the LabelOverlapMeasuresImageFilter submodule,
so SWIG had only a forward declaration and could not materialize the
hashmap wrapper class.  Confirmed locally on macOS: count of `hashmap`
in the SWIG-generated ITKImageStatisticsPython.cpp was 0 across every
namespace-qualifier and ordering variant tried.

Drop the unused infrastructure so the PR ships only the working
C++/Python surface introduced in the previous commit:

  * Remove WRAPPER_SWIG_LIBRARY_FILES wiring from
    Modules/Filtering/ImageStatistics/wrapping/CMakeLists.txt.
  * Remove itkLabelOverlapLabelSetMeasuresInstantiations.i (no
    longer referenced).
  * Revert the `%include <std_unordered_map.i>` line added to
    Wrapping/Generators/Python/PyBase/pyBase.i (unused after the
    above; pyBase.i is foundation-level shared infrastructure and
    should not carry a header it doesn't consume).

Update itkLabelOverlapMeasuresImageFilterTest.py to use the new
GetLabels()/GetMeasureForLabel() accessors and the explicit getters
(GetIntersection, GetUnion) on the per-label struct.  Local test
output (200 voxel-label fixture, 14 labels detected, intersection
and union counts match expected per-label values):

  Found 14 labels
  Label: 0, ...  Label: 1, i: 2246, u: 120691  ...  Label: 13, i: 0, u: 387126
  100% tests passed, 0 tests failed out of 1

Co-Authored-By: dzenanz <dzenanz@gmail.com>

hjmjohnson commented:
Pushed two commits on top of `lsmLabelOverlap` that take a different angle on the Python-side test failure.

**Diagnosis (verified locally, macOS Python wrapping build):** SWIG silently drops `%template std::unordered_map<X, itk::LabelOverlapLabelSetMeasures>` in the `Instantiations.i` because the value type is `%import`-ed (not `%include`-d) into the consuming submodule. Count of `hashmap` in the SWIG-generated `ITKImageStatisticsPython.cpp` was 0 across every namespace-qualifier and ordering variant tried. Confirmed both inside `Modules/Filtering/ImageStatistics/wrapping/` and after relocating the `%template` declarations to `pyBase.i` — the second attempt failed for layering reasons (pyBase.i is foundation and can't see leaf-module headers).

**Pivot to a wrapping-friendly C++ surface:** added two small accessors that don't rely on `std::unordered_map` getting a Python-dict wrapper. The existing `MapType GetLabelSetMeasures()` is preserved unchanged for C++ consumers.

```cpp
std::vector<LabelType>      GetLabels() const;
LabelOverlapLabelSetMeasures GetMeasureForLabel(LabelType) const;
```

Plus six `Get*` getters on the struct itself (igenerator doesn't auto-wrap public data members). Both work natively across the per-submodule SWIG isolation because `std::vector<unsigned char>` is already wrapped globally in `pyBase.i` and the struct is wrapped via the existing `itk_wrap_simple_class` directive.

**Test now passes locally** against the existing fixture:

```
Found 14 labels
Label: 1, i: 2246, u: 120691
Label: 2, i: 497,  u: 22835
…
Label: 13, i: 0,   u: 387126
1/1 Test #211: LabelOverlapMeasuresImageFilterTest ...   Passed    1.75 sec
100% tests passed, 0 tests failed out of 1
```

<details>
<summary>Commits (2, on top of your existing 2)</summary>

| | Commit | Files |
|---|---|---|
| 3 | `ENH: Add Python-friendly LabelSetMeasures accessors` | `itkLabelOverlapMeasuresImageFilter.h`, `itkLabelOverlapLabelSetMeasures.h` |
| 4 | `COMP: Drop unused std::unordered_map SWIG infrastructure` | `wrapping/CMakeLists.txt` (drop `WRAPPER_SWIG_LIBRARY_FILES`), delete `itkLabelOverlapLabelSetMeasuresInstantiations.i`, revert `pyBase.i` `%include <std_unordered_map.i>`, update Python test |

Both commits carry `Co-Authored-By: dzenanz` to preserve your authorship.

</details>

<details>
<summary>What I tried before pivoting (so this isn't a black box)</summary>

| Variant | Result |
|---|---|
| Original (`itk::LabelOverlapLabelSetMeasures` qualified, in `Instantiations.i`) | `%template` silently dropped — 0 hashmap classes generated |
| Drop the `itk::` namespace qualifier | SWIG generates templates but C++ can't compile (the flat name only exists as a typedef in the struct's submodule) |
| Add `using itkLabelOverlapLabelSetMeasures = itk::LabelOverlapLabelSetMeasures;` to `Instantiations.i`'s `%{…%}` | Templates still dropped |
| Move `%template` to `pyBase.i` with `%import "itkLabelOverlapLabelSetMeasures.h"` | SWIG fails: header not on PyBase's include path |
| Forward-declare in `pyBase.i` (`namespace itk { struct LabelOverlapLabelSetMeasures; }`) | SWIG generates code; C++ compile fails for layering reasons |

Root cause is architectural: ITK's wrapping pipeline gives each submodule its own SWIG run with `%import`-only type info from siblings, and SWIG's `%template std::unordered_map<X, Y>` requires Y to be `%include`-d (full type definition) — not just declared. Fixing this cleanly would be a multi-day igenerator.py change with broad blast radius.

</details>

CI will re-run on `2e882122e2`. If anything in this approach doesn't sit right with you, happy to iterate — the commits are surgical and easy to drop / amend.

hjmjohnson commented:
Re-triggering ARMBUILD-x86_64-rosetta — `itkImageAlgorithmCopyTest2` (3.14 float CheckBuffer) failed in isolation on this runner; same test passed on #6185/#6188/#6189 against the same runner pool, and CDash shows 0 test failures (Linux 3486/3486, Darwin 309/309, Linux.Python 176/176). Known x86_64-rosetta float-precision flake, unrelated to this PR's diff.

dzenanz reviewed (commented):
Looks good. I can't approve my own PR.

hjmjohnson commented on Modules/Filtering/ImageStatistics/include/itkLabelOverlapMeasuresImageFilter.h:
```diff
@@ -71,25 +71,14 @@ class ITK_TEMPLATE_EXPORT LabelOverlapMeasuresImageFilter : public ImageSink<TLa
-    // default constructor/copy/move etc...
+#ifndef ITK_FUTURE_LEGACY_REMOVE
+  /** Type to use for computations. */
+  using LabelSetMeasures = LabelOverlapLabelSetMeasures;
```
Addressed in e04e97e576 — comment now reads `/** Deprecated backward-compatibility alias.  Use LabelOverlapLabelSetMeasures directly. */`, no longer copy-pasted from RealType.

hjmjohnson commented on Modules/Filtering/ImageStatistics/include/itkLabelOverlapMeasuresImageFilter.h:
```diff
@@ -71,25 +71,14 @@ class ITK_TEMPLATE_EXPORT LabelOverlapMeasuresImageFilter : public ImageSink<TLa
-    SizeValueType m_Intersection{ 0 };
-    SizeValueType m_SourceComplement{ 0 };
-    SizeValueType m_TargetComplement{ 0 };
-  };
 
   /** Type of the map used to store data per label */
```
Addressed in e04e97e576 — collapsed the double blank line between the `#endif` guard and the next declaration to a single blank, matching ITK style.

hjmjohnson commented on Modules/Filtering/ImageStatistics/wrapping/test/itkLabelOverlapMeasuresImageFilterTest.py:
```diff
@@ -0,0 +1,42 @@
+    measure = lom_filter.GetMeasureForLabel(label)
+    # Use the explicit Get* accessors (igenerator does not expose public data
+    # members of a struct to Python; the m_Foo fields are not visible).
+    print(f"Label: {label}, i: {measure.GetIntersection()}, u: {measure.GetUnion()}")
```
Addressed in e04e97e576 — added three assertions:
- `assert len(labels) > 0` (filter must produce at least one label)
- `assert intersection >= 0` (sanity: counts non-negative)
- `assert union >= intersection` (set-theoretic invariant: |A ∪ B| ≥ |A ∩ B|)

A wrong Dice/Jaccard would now be caught by the union-vs-intersection check; a regression that drops all labels would be caught by the first assertion.
